src/app/component/ProductReviewForm/ProductReviewForm.component.js

```diff
@@ -0,0 +1,253 @@
+ */
+
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import { ProductType } from 'Type/ProductList';
+import { RatingItemsType } from 'Type/Rating';
+import Field from 'Component/Field';
+import Form from 'Component/Form';
+import ContentWrapper from 'Component/ContentWrapper';
+import TextPlaceholder from 'Component/TextPlaceholder';
+import Loader from 'Component/Loader';
+import { customerType } from 'Type/Account';
+import './ProductReviewForm.style';
+
+/**
+ * ProductReviewForm
+ * @class ProductReviewForm
+ */
+class ProductReviewForm extends Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            isLoading: false,
+            ratingData: []
+        };
+    }
+
+    onReviewSubmitAttempt(fields, invalidFields) {
+        this.setState({ isLoading: true });
+        const { showNotification } = this.props;
+
+        if (invalidFields) {
+            showNotification('error', 'Incorrect data! Please resolve all field validation errors.');
+        }
+
+        this.setState({ isLoading: !invalidFields });
+    }
+
+    onReviewSubmitSuccess(fields) {
+        const {
+            product,
+            addReview,
+            showNotification
+        } = this.props;
+
+        const {
+            nickname,
+            title,
+            detail
+        } = fields;
+
+        const { sku } = product;
+        const { ratingData } = this.state;
+        const ratingsAreValid = this.validateRatings();
+
+        this.setState({ isLoading: true });
+
+        if (!ratingsAreValid) {
+            showNotification('error', 'Incorrect rating data! Please make sure that all ratings are selected.');
+            this.setState({ isLoading: false });
+        } else {
+            addReview({
+                nickname,
+                title,
+                detail,
+                product_sku: sku,
+                rating_data: ratingData
+            }).then((success) => {
+                /** Clear form after submitting valid review */
+                if (success) {
+                    this.formRef.form.reset();
+                    this.setState({ ratingData: [] });
+                }
+
+                this.setState({ isLoading: false });
+            });
+        }
+    }
+
+    validateRatings() {
+        const { reviewRatings } = this.props;
+        const { ratingData } = this.state;
+        let isValid = true;
+
+        reviewRatings.forEach((activeRating) => {
+            if (!ratingData.some(selectedRating => selectedRating.rating_id === activeRating.rating_id)) {
+                isValid = false;
+            }
+        });
+
+        return isValid;
+    }
+
+    handleReviewRatingChange(ratingId, optionId) {
+        if (typeof ratingId === 'number' && typeof optionId === 'number') {
+            const { ratingData } = this.state;
+
+            if (ratingData.length > 0 && ratingData.some(rating => rating.rating_id === ratingId)) {
```

Comment:
Maybe if you are often searching in `raitingData` it makes sense to prepare data in reducer in the correct way? For example, use `rating_id` as a key. For example from your array the object with raiting_id keys should be prepared. THen there will be no need to go through the raiting every time, find raiting it...

Reply:
By this you mean that `ratingData` should be stored with redux? Because currently `ratingData` initially is just an empty state array that gets filled with `{ rating_id: ..., option_id: ... }` objects when the ratings get selected and this array then gets used to save selected rating data upon form submission. Maybe I misunderstood something.

Reply:
Just use key => value instead of simple pushing into array.